docs/design/rosbag2_playback_time.md

```diff
@@ -0,0 +1,115 @@
+
+In the final case, when rosbag2 is driven by an external source, it still must be able to handle the above operations, which may be performed by that external time controller. Therefore, a generic solution is required, so that  `rosbag2_transport::Player` is agnostic to the controller of the time.
+
+## Current Implementation Notes
+
+`rosbag2_transport::Player` currently uses `std::chrono::system_clock` to query time, and `std::this_thread::sleep_until` to wait between publishing messages, with explicit handling of playback rate.
+
+## Proposal
+
+Pass a `rclcpp::Clock` instance to `rosbag2_transport::Player`
+* Use `Clock::now` to query starting time
+* Use `Clock::sleep_until` between messages
+  * `rclcpp::Clock` does not yet implement `sleep_until` - as noted in https://github.com/ros2/rcl/issues/898
+  * If we are not able to get into `rclcpp` in time for the Galactic API freeze - it may need to temporarily go into a rosbag2-based subclass for the Galactic release.
+
+When an external source drives time, a plain `rclcpp::Clock` can be used, configured to ROStime as the source. For situations where rosbag2 drives time, a `rosbag2_transport::PlayerClock` class is created - which returns the correct time based on requirements (see API declaration below).
+
+```
+class PlayerClock : public rclcpp::Clock
+{
+public:
+  /**
+   * playback_rate must be positive, 1.0 means real-time
+   * starting_time provides an offset for all future times
+   * clock_topic_publish_frequency:
+   *   - if > 0 - publishes `now()` to /clock at the specified frequency in Hz, rate determined by a system steady clock
+   *   - if <= 0 - does not publish the /clock topic
+   */
+  Clock(float playback_rate, rclcpp::Time starting_time, float clock_topic_publish_frequency);
+
+  // rclcpp::Clock interface
+  /**
+   * based on an internal steady clock, calculates current "Player Time"
+   */
+  rclcpp::Time now() override;
+
+  /**
+   * Sleeps (non-busy-waiting) the current thread until the provided time is reached - according to this Clock!
+   *
+   * If time is paused, the requested time may never be reached
+   * `real_time_timeout` uses the internal steady clock, if the timeout elapses, return false
+   * If jump() is called, return false, allowing the caller to handle the new time
+   * Return true if the time has been reached
+   */
+  bool sleep_until(rclcpp::Time until, rclcpp::Duration real_time_timeout) override;
+
+  // Special playback interface
+
+  /**
+   * Pauses/resumes time.
+   * While paused, `now()` will repeatedly return the same time, until resumed
+   *
+   * Note: this could have been implemented as `set_rate(0)`, but this interface allows this clock to maintain the
+   * clock's rate internally, so that the caller does not have to save it in order to resume.
+   */
+  void set_paused(bool paused);
+
+  /**
+    * Changes the rate of playback. Must be positive (nonzero)
+    */
+  void set_rate(float rate);
```

Comment:
Implementation detail, but should this `throw` if `rate < zero`?

Reply:
I assumed that's what i would do to enforce the "Must be positive nonzero" requirement